### ISSUE TITLE
NVDA doesn't read colomn and row headers in Libreoffice calc tables

### ISSUE THREAD

**Issue by pavel46** [nvaccess/nvda#17274]
#### NVDA installed/portable/running from source:
installed
#### NVDA version:
2024.3.1
#### Windows version:
windows 11 23H2
#### Name and version of other software in use when reproducing the issue:
Libreoffice Calc 24/8
#### Does the issue still occur after restarting your computer?
yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
no
#### If NVDA add-ons are disabled, is your problem still occurring?
yes

**Comment by Adriani90** [comment]
cc: @michaelweghorn are these exposed correctly via iAccessible in Calc? Or should we create a bug in the LO bug tracker for that?

**Comment by michaelweghorn** [comment]
> hi all, simulare to the situation with the tables in LibreOffice Writer NVDA doesn't read colomn and row headers in the LibreOffice calc. i usually define cells in the first row as a colomn headers and as far as i can see with the rest of my vision capibilites LibreOffice calc is define those cells as colomn headers, because when i go to the view menu and set highliting colomn and row headers the first cell of the current colomn is highlited with the different color. 

Can you please clarify the exact name of the option in the "View" menu you are enabling? It's currently not clear to me yet which of them would result in auto-detection of what is a row or column header.

> cc: @michaelweghorn are these exposed correctly via iAccessible in Calc? Or should we create a bug in the LO bug tracker for that?

One of the questions would be how to reliably determine what a row/column header is, as it's in theory possible to have multiple custom "tables" inside of a Calc spreadsheet, or multiple headers,...

Issue #15394 might be related.

S.a. my comment about Writer tables in https://github.com/nvaccess/nvda/issues/15394#issuecomment-1710757471 , where a particular paragraph style seems to be used to detect headings.

[The NVDA user guide section for Word row and column headers](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#WordAutomaticColumnAndRowHeaderReading) mentions this:

> NVDA is able to automatically announce appropriate row and column headers when navigating around tables in Microsoft Word. This requires that the Report Table row / column headers option in NVDA's Document Formatting settings, found in the [NVDA Settings](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#NVDASettings) dialog, be turned on.
> 
> If you use [UIA to access Word documents](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#MSWordUIA), which is default in recent versions of Word and Windows, the cells of the first row will automatically be considered as column headers; similarly, the cells of the first column will automatically be considered as row headers.
>
> On the contrary, if you do not use [UIA to access Word documents](https://www.nvaccess.org/files/nvda/documentation/userGuide.html#MSWordUIA), you will have to indicate to NVDA which row or column contains the headers in any given table. After moving to the first cell in the column or row containing the headers, use one of the following commands:
>
> [...]
>
> These settings will be stored in the document as bookmarks compatible with other screen readers such as JAWS. This means that users of other screen readers who open this document at a later date will automatically have the row and column headers already set.

I'm not familiar with these, but it sounds to me like that would be logic (that would be) implemented on the screen reader side rather than in MS Office/LibreOffice. Is that correct?

**Comment by Adriani90** [comment]
@michaelweghorn in general in Microsoft  Excel you can define row and column headers by defining a name range from the selected cells.
e.g.
1. Select a part of a sheet (e.g. cells a5:d9)
2. Press ctrl+shift+f3 to open the "create named range from selection dialog"
3. Select which row and column should be the hader row and header column from the selected range of cells (e.g. first row,, first column, last row etc.)
4. Press ok.

At this point, the selected rows and columns in the named range are now set as row and column headers.

NVDA treats these and reports them as headers because Excel exposes them as named rows and columns. This is defined in the object window for MS Excel oin NVDA.
NVDA provides a very comfortable key stroke (nvda+shift+c for column, and NVDA+shift+r for row headers). These key strokes actually use the "define name range" feature in Excel as per the steps above, beginning with the cell  you pressed these key strokes.
Supose I press nvda+shift+r in cell b3, then the cells b3 until next empty cell are defined as named range as per steps above in MS Excel. So the header row starts in cell b3 and goes until the next empty cell on the same row.
The same occurs for column headers but downwards.

So in summary, if Libre Office calc would allow setting row and column headers via defining named ranges, then NVDA could handle these via the SOffice appmodule, in case the same approach is applied as in MS Excel.

The support for row and column headers for Excel is implemented in 
https://github.com/nvaccess/nvda/blob/a1f9423ccdfc639a7edb8c02420ef763d144d711/source/NVDAObjects/window/excel.py

I think starting at about line 600 or so.


The same approach could be applied for tables created in Libre Office writer, in case named ranges can be defined there as well.
In contrast, in MS Word only one row and one column per table can be defined as header, and this is what NVDA recognizes as such. In general it is the first row and the first column. This can be defined in the table properties in MS Word.

**Comment by Adriani90** [comment]
See also e.g. #6629 for inspiration on how NVDA handles headers in MS Excel.
